File: backend/src/Ergoplanner.Domain/Entities/SymbolMetadata.cs
using System;
using System.Collections.Generic;
using Ergoplanner.Domain.Common;
using Ergoplanner.Domain.Interfaces;

namespace Ergoplanner.Domain.Entities;

public class SymbolMetadata : BaseEntity, ISymbolMetadata
{
    public SymbolMetadata()
    {
        CustomProperties = new Dictionary<string, object>();
        Version = 1;
    }

    // Identification Properties
    public string TagNumber { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string Category { get; set; } = string.Empty;
    public string SubCategory { get; set; } = string.Empty;

    // Technical Specifications
    public string? Size { get; set; }
    public string? Rating { get; set; }
    public string? Material { get; set; }
    public string? Type { get; set; }
    public string? Model { get; set; }
    public string? Manufacturer { get; set; }

    // Process Data
    public string? Service { get; set; }
    public decimal? DesignPressure { get; set; }
    public string? DesignPressureUnit { get; set; }
    public decimal? DesignTemperature { get; set; }
    public string? DesignTemperatureUnit { get; set; }
    public decimal? FlowRate { get; set; }
    public string? FlowRateUnit { get; set; }
    public decimal? OperatingPressure { get; set; }
    public string? OperatingPressureUnit { get; set; }
    public decimal? OperatingTemperature { get; set; }
    public string? OperatingTemperatureUnit { get; set; }

    // Standards Compliance
    public string? ISAStandard { get; set; }
    public string? PIPStandard { get; set; }
    public string? ISOStandard { get; set; }
    public string? DINStandard { get; set; }
    public string? BSStandard { get; set; }

    // Additional Properties (stored as JSON)
    public Dictionary<string, object> CustomProperties { get; set; }

    // Relationships
    public Guid SymbolId { get; set; }
    public virtual Symbol? Symbol { get; set; }

    public Guid? ParentMetadataId { get; set; }
    public virtual SymbolMetadata? ParentMetadata { get; set; }

    public string? TemplateId { get; set; }
    public virtual ICollection<SymbolMetadata>? ChildMetadata { get; set; }

    // Audit Properties
    public DateTime? LastModifiedAt { get; set; }
    public string? LastModifiedBy { get; set; }
    public int Version { get; set; }

    // Methods for property inheritance
    public void InheritFrom(SymbolMetadata parent)
    {
        if (parent == null) return;

        // Inherit technical specs if not set
        Size ??= parent.Size;
        Rating ??= parent.Rating;
        Material ??= parent.Material;
        Type ??= parent.Type;

        // Inherit process data if not set
        Service ??= parent.Service;
        DesignPressure ??= parent.DesignPressure;
        DesignPressureUnit ??= parent.DesignPressureUnit;
        DesignTemperature ??= parent.DesignTemperature;
        DesignTemperatureUnit ??= parent.DesignTemperatureUnit;

        // Inherit custom properties that don't exist in child
        foreach (var prop in parent.CustomProperties)
        {
            if (!CustomProperties.ContainsKey(prop.Key))
            {
                CustomProperties[prop.Key] = prop.Value;
            }
        }
    }

    public SymbolMetadata Clone()
    {
        return new SymbolMetadata
        {
            TagNumber = TagNumber,
            Name = Name,
            Description = Description,
            Category = Category,
            SubCategory = SubCategory,
            Size = Size,
            Rating = Rating,
            Material = Material,
            Type = Type,
            Model = Model,
            Manufacturer = Manufacturer,
            Service = Service,
            DesignPressure = DesignPressure,
            DesignPressureUnit = DesignPressureUnit,
            DesignTemperature = DesignTemperature,
            DesignTemperatureUnit = DesignTemperatureUnit,
            FlowRate = FlowRate,
            FlowRateUnit = FlowRateUnit,
            OperatingPressure = OperatingPressure,
            OperatingPressureUnit = OperatingPressureUnit,
            OperatingTemperature = OperatingTemperature,
            OperatingTemperatureUnit = OperatingTemperatureUnit,
            ISAStandard = ISAStandard,
            PIPStandard = PIPStandard,
            ISOStandard = ISOStandard,
            DINStandard = DINStandard,
            BSStandard = BSStandard,
            CustomProperties = new Dictionary<string, object>(CustomProperties),
            TemplateId = TemplateId,
            Version = 1
        };
    }
}

File: frontend/src/types/symbolMetadata.ts
/**
 * Symbol Metadata types matching the backend API
 */

export interface SymbolMetadata {
  id: string;
  symbolId: string;
  tagNumber: string;
  name: string;
  description: string;
  category: string;
  subCategory: string;
  
  // Technical Specifications
  size?: string;
  rating?: string;
  material?: string;
  type?: string;
  model?: string;
  manufacturer?: string;
  
  // Process Data
  service?: string;
  designPressure?: number;
  designPressureUnit?: string;
  designTemperature?: number;
  designTemperatureUnit?: string;
  flowRate?: number;
  flowRateUnit?: string;
  
  // Standards
  isaStandard?: string;
  pipStandard?: string;
  isoStandard?: string;
  dinStandard?: string;
  bsStandard?: string;
  
  // Additional Properties
  customProperties: Record<string, any>;
  
  // Audit Fields
  createdAt: string;
  createdBy: string;
  lastModifiedAt?: string;
  lastModifiedBy?: string;
  version: number;
  
  // Relationships
  parentMetadataId?: string;
  childMetadataIds?: string[];
  templateId?: string;
}

export interface PropertyTemplate {
  id: string;
  name: string;
  category: string;
  description?: string;
  requiredProperties: string[];
  optionalProperties: string[];
  defaultValues: Record<string, any>;
  validationRules?: Record<string, any>;
  isActive: boolean;
  createdAt: string;
  createdBy: string;
}

export interface CreateSymbolMetadataRequest {
  symbolId: string;
  tagNumber: string;
  name: string;
  description: string;
  category: string;
  subCategory: string;
  size?: string;
  rating?: string;
  material?: string;
  type?: string;
  model?: string;
  manufacturer?: string;
  service?: string;
  designPressure?: number;
  designPressureUnit?: string;
  designTemperature?: number;
  designTemperatureUnit?: string;
  flowRate?: number;
  flowRateUnit?: string;
  isaStandard?: string;
  pipStandard?: string;
  isoStandard?: string;
  customProperties?: Record<string, any>;
  parentMetadataId?: string;
  templateId?: string;
}

export interface UpdateSymbolMetadataRequest {
  id: string;
  tagNumber: string;
  name: string;
  description: string;
  size?: string;
  rating?: string;
  material?: string;
  type?: string;
  service?: string;
  designPressure?: number;
  designPressureUnit?: string;
  designTemperature?: number;
  designTemperatureUnit?: string;
  customProperties?: Record<string, any>;
}

export const METADATA_CATEGORIES = [
  'Process Equipment',
  'Piping Components', 
  'Instrumentation',
  'Electrical',
  'Safety',
] as const;

export const PRESSURE_UNITS = ['bar', 'psi', 'kPa', 'MPa', 'atm'] as const;
export const TEMPERATURE_UNITS = ['°C', '°F', 'K'] as const;
export const FLOW_UNITS = ['m³/h', 'gpm', 'L/min', 'L/s', 'ft³/h'] as const;

export type MetadataCategory = typeof METADATA_CATEGORIES[number];
export type PressureUnit = typeof PRESSURE_UNITS[number];
export type TemperatureUnit = typeof TEMPERATURE_UNITS[number];
export type FlowUnit = typeof FLOW_UNITS[number];